red/runtime/nodes/context/localfilesystem.js

```diff
@@ -0,0 +1,156 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+var fs = require('fs-extra');
+var path = require("path");
+var when = require("when");
+var util = require("../../util");
+
+function getStoragePath(storageBaseDir, scope) {
+    if(scope.indexOf(":") === -1){
+        if(scope === "global"){
+            return path.join(storageBaseDir,"global",scope);
+        }else{ // scope:flow
+            return path.join(storageBaseDir,scope,"flow");
+        }
+    }else{ // scope:local
+        var ids = scope.split(":")
+        return path.join(storageBaseDir,ids[1],ids[0]);
+    }
+}
+
+function getBasePath(config) {
+    var base = config.base || "contexts";
+    var storageBaseDir;
+    if (!config.dir) {
+        if(config.settings && config.settings.userDir){
+            storageBaseDir = path.join(config.settings.userDir, base);
+        }else{
+            try {
+                fs.statSync(path.join(process.env.NODE_RED_HOME,".config.json"));
+                storageBaseDir = path.join(process.env.NODE_RED_HOME, base);
+            } catch(err) {
+                try {
+                    // Consider compatibility for older versions
+                    if (process.env.HOMEPATH) {
+                        fs.statSync(path.join(process.env.HOMEPATH,".node-red",".config.json"));
+                        storageBaseDir = path.join(process.env.HOMEPATH, ".node-red", base);
+                    }
+                } catch(err) {
+                }
+                if (!storageBaseDir) {
+                    storageBaseDir = path.join(process.env.HOME || process.env.USERPROFILE || process.env.HOMEPATH || process.env.NODE_RED_HOME,".node-red", base);
+                }
+            }
+        }
+    }else{
+        storageBaseDir = path.join(config.dir, base);
+    }
+    return storageBaseDir;
+}
+
+function loadFile(storagePath){
+    return fs.pathExists(storagePath).then(function(exists){
+        if(exists === true){
+            return fs.readFile(storagePath, "utf8");
+        }else{
+            return when.resolve(undefined);
+        }
+    }).catch(function(err){
+        throw when.reject(err);
+    });
+}
+
+function LocalFileSystem(config){
+    this.config = config;
+    this.storageBaseDir = getBasePath(this.config);
+}
+
+LocalFileSystem.prototype.open = function(){
+    return when.resolve();
+}
+
+LocalFileSystem.prototype.close = function(){
+    return when.resolve();
+}
+
+LocalFileSystem.prototype.getAsync = function(scope, key) {
```

Comment:
⚠️ My understanding is the context storage plugins are only ever accessed with the async get/set functions - I don't think we should have both `get/set` or `getAsync/setAsync` here (or with the other plugins)